### DISCUSSION TITLE
Load plugin before all other plugins

### DISCUSSION THREAD

**Discussion by robsonpeixoto** [LazyVim/LazyVim#1389] [Q&A] [answered]
I'm trying to configure a better helm support, but important to detect the filetype before load any plugin, to run only the language servers that makes sense.

I tried to set a very high priority, but some language servers that should not be running, is running:

The filetype is right:
![image](https://github.com/LazyVim/LazyVim/assets/124390/263927fb-4c54-405a-94bf-b7c2332ba60c)

But some language servers is running for `yaml` file type:
```
 Press q or <Esc> to close this window. Press <Tab> to view server doc.
 
 Language client log: /home/my-user/.local/state/nvim/lsp.log
 Detected filetype:   helm
 
 3 client(s) attached to this buffer: 
 
 Client: null-ls (id: 1, bufnr: [3])
 	filetypes:       fish, luau, lua, sh, javascriptreact, handlebars, graphql, yaml, markdown.mdx, typescript, vue, json, less, markdown, jsonc, typescriptreact, scss, html, javascript, css, dockerfile, gitcommit, go, python, tf, terraform, terraform-vars, toml
 	autostart:       false
 	root directory:  /home/my-user/tmp/examples
 	cmd:             <function>
 
 Client: docker_compose_language_service (id: 2, bufnr: [3])
 	filetypes:       yaml
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /home/my-user/.local/share/nvim/mason/bin/docker-compose-langserver --stdio
 
 Client: helm_ls (id: 3, bufnr: [3])
 	filetypes:       helm
 	autostart:       true
 	root directory:  /home/my-user/tmp/examples/charts/hello-world
 	cmd:             /home/my-user/.local/share/nvim/mason/bin/helm_ls serve
 
 Configured servers list: docker_compose_language_service, dockerls, lua_ls, helm_ls, jsonls, pyright, erlangls, tsserver, eslint, gopls, bashls, terraformls, rust_analyzer, tflint, ruff_lsp, taplo, elixirls
```

Any idea of how to solve it?

My `helm.lua` plugin:
```lua
return {
  {
    "towolf/vim-helm",
    lazy = false,
    priority = 900,
  },
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      servers = {
        helm_ls = {},
      },
    },
  },
}
```

References:
- https://github.com/LazyVim/LazyVim/pull/1218
- https://github.com/LazyVim/LazyVim/discussions/185

**Comment by abeldekat** [comment]
Hello @robsonpeixoto,

The main problem is the configuration of the docker_compose_language_service, triggering on each `yaml` file.

See this PR on the [`nvim-lspconfig plugin`](https://github.com/neovim/nvim-lspconfig/pull/2772)
A quote:

> I believe the original intention was to have all Docker-Compose files be recognized and have features as a FileType of YAML, but in reality the inverse was configured (all YAML files are assumed to be Docker-Compose files).

A quick solution to mitigate this: 
Start the docker_compose_language_service manually, using `:LspStart docker_compose_language_service`

```lua
return {
  {
    "towolf/vim-helm",
    -- lazy = false, -- Defaults to lazy = false
    -- priority = 900, -- The priority is not relevant here
  },
  {
    "nvim-lspconfig",
    opts = {
      servers = {
        docker_compose_language_service = {
          autostart = false,
        },
        helm_ls = {},
      },
    },
  },
}
```

When opening a `yaml` file for `helm`, the filetype is set in two steps:

- `neovim` filetype detection: set filetype to yaml
- `vim-helm` plugin, ftdetect: set filetype to helm

The final filetype is `helm`.
Both steps trigger the `autocommands` the `nvim-lspconfig` plugin has registered on the `FileType event`, resulting in the activation of both `lsp`'s.

Thus, if you want to keep the docker_compose_language_service on `autostart`, then the `vim-helm` plugin is not sufficient.

Solution: Augment `neovim`'s filetype detection(`:h vim.filetype.add`), effectively removing the second step.
The following code snippet is a modified version of `AstroNvim`'s [ansible-recipe](https://astronvim.com/Recipes/yaml_ansible)

```lua
local function yaml_ft(path, buffer)
  if (rewrite_the_criteria_of_the_vim-helm_plugin) then
    return "helm"
  end
  return "yaml"
end

return {
  {
    "nvim-lspconfig",
    init = function()
      vim.filetype.add({
        extension = {
          yml = yaml_ft,
          yaml = yaml_ft,
        },
      })
    end,
    opts = {
      servers = {
        docker_compose_language_service = {},
        helm_ls = {},
      },
    },
  },
}
```

Best regards!

**Comment by robsonpeixoto** [comment]
Thanks @abeldekat. Was not necessary to disable `docker_compose_language_service`, only the `vim.api.nvim_create_autocmd` was enough.

```lua
-- hack to define filetype=helm before all plugins be loaded
vim.api.nvim_create_autocmd({ "BufNewFile", "BufRead" }, {
  pattern = {
    "*/templates/*.yaml",
    "*/templates/*.tpl",
    "*.gotmpl",
    "helmfile*.yaml",
  },
  callback = function()
    vim.bo.filetype = "helm"
    vim.bo.commentstring = "{{/* %s */}}"
  end,
})

return {
  { "towolf/vim-helm" },
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      servers = {
        helm_ls = {},
      },
    },
  },
}
```

**Reply by abeldekat** [reply, marked as answer]
Hello @robsonpeixoto,

Your solution works, however, in my opinion, there are some very minor technical drawbacks to note:

- An `autocommand` on `FileType` is put _before_ `neovim`'s `filetype` detection. **Correction**: An `autocommand` on `BufNewFile and BufRead` is put _before_ `neovim`'s `filetype` detection. Thanks, @dpetka2001
  As a consequence, the patterns in that `autocommand` will always be ran,
  also on other types.
- The `autocommand` replicates the code from the `vim-helm` plugin.
  Still, the `autocommands` from the `vim-helm` plugin are also registered.

The ideal solution, in my opinion, has the following traits:

- No code needs to be duplicated when using a plugin like `vim-helm`, containing both `ftdetect` and `syntax`.
- Targets the core problem: The `docker compose lsp` does not have a dedicated `filetype`, thus starts on each `yaml`.

For example, add a file `lua/plugins/lsp.lua`:

```lua
require("lazyvim.util").on_attach(function(client, buffer)
  if client.name == "docker_compose_language_service" then
    if vim.api.nvim_buf_get_option(buffer, "filetype") == "helm" then
      vim.schedule(function()
        vim.cmd("LspStop ++force docker_compose_language_service")
      end)
    end
  end
end)

return {
  "towolf/vim-helm",
  {
    "nvim-lspconfig",
    opts = {
      servers = {
        docker_compose_language_service = {},
        helm_ls = {},
      },
    },
  },
}
```
